sw/host/opentitanlib/src/io/spi.rs

```diff
@@ -151,6 +151,39 @@ pub trait Target {
         Err(SpiError::InvalidOption("This target does not support set_voltage".to_string()).into())
     }
 
-    /// Runs a SPI transaction composed from the slice of [`Transfer`] objects.
+    /// Runs a SPI transaction composed from the slice of [`Transfer`] objects.  Will assert the
+    /// CS for the duration of the entire transactions.
     fn run_transaction(&self, transaction: &mut [Transfer]) -> Result<()>;
+
+    /// Assert the CS signal.  Uses reference counting, will be deasserted when each and every
+    /// returned `AssertChipSelect` object have gone out of scope.
+    fn assert_cs<'a>(&'a self) -> Result<AssertChipSelect<'a>> {
+        Err(SpiError::InvalidOption("This target does not support assert_cs".to_string()).into())
+    }
```

Comment:
How is this going to interact with the network proxy transport?

Reply:
We will need new protocol requests to assert and deassert the CS (increment/decrement the count, really).  This method will be implemented in the client library by sending an increment request, and returning an object which when dropped sends the decrement request.  On the network server end, we probably are going to have a vector of AssertChipSelect objects associated with each connection, handling the increment request by invoking assert_cs on the underlying transport and putting its result on the vector, and handling decrement requests by removing an element and dropping it.  (If the connection is broken, the stack of objects will also be dropped, releasing any chip select on the underlying transport.)

In short, I think it can be done.  (One could avoid the vector in favor of an Option<AssertChipSelect> and a separate integer, to have the proxy server do the counting.)

Reply:
My problem now is that in TransportCommandHandler.do_execute_cmd() we have a &'a TransportWrapper on which a call to spi() is made, returning a Rc<dyn spi::Target>.  We want to invoke the new assert_cs() on the spi::Target borrowed from that Rc, but for all the compiler knows all the Rc objects could run out of scope before the AssertChipSelect, if we had put in on a vector.
In order to switch to runtime lifetime management, the AssertChipSelect would probably need to hold a Rc<dyn spi::Target>, rather than a direct reference to the object.  This could possibly be done by modifying the signature into:
```
fn assert_cs(self: Rc<dyn Target>) -> Result<AssertChipSelect>;
```
I am not sure if it is possible to create methods this way, though.

Reply:
There are a few forms of `self` allowed.  Basically, any "smart pointer" formulation of `self` is permitted:
```
fn assert_cs(self: &Rc<Self>) -> Result<foo>
```

Reply:
I will try this.  However, my original idea was to not have the deassert() method be part of the spi::Target trait, and instead have the common AssertChipSelect struct invoke deasser() through a second trait.  Any struct implementing spi::Target would also implment my second "hidden" trait, TargetChipDeassert, and that way be able to construct a AssertChipSelect passing its self cast into a dyn TargetChipDeassert.  When I instead have a Rc<dyn Target> or Rc of some specific implementing struct, I do not think that I can similarly upcast to Rc<dyn TargetChipDeassert>.

I will see what I can do, but I am afraid that I will not be able to find a way for AssertChipSelect to hold a Rc<> reference to one out of various implementation spi::Target, in such a way that it can invoke a method that is not exposed as part of the trait.

Reply:
The only thorn is that Rust apparently does not support `self: &Rc<Self>`, but only supports the non-reference version, in order to be "object safe".  The description in the link gives me the impression that it is not an inherent limitation of the type system, but rather a matter of it not being implement, yet.  We can seamlessly remedy this in the future, if the Rust compile should get support, for now, the caller will have to clone the `Rc` ahead of calling `assert_cs()`.  I think that is acceptable.  
https://doc.rust-lang.org/reference/items/traits.html#object-safety